File: src/tui/graph/glyphs.ts
/**
 * Graph gutter glyphs — Unicode (lazygit-class) or ASCII (`git log --graph`).
 *
 * Node semantics (Unicode): commit/merge `●`, HEAD `⊙`, uncommitted `○`,
 * stash `◇`. Merge topology (not a distinct node glyph) identifies merges.
 */

import { asciiGlyphs } from '../icons.js';

/** Columns occupied per logical lane (glyph + spacer / horizontal). */
export const CELL_W = 2;

/**
 * Glyph set for one paint mode.
 */
export type GraphGlyphSet = {
  readonly commit: string;
  /** HEAD commit — including merge-at-HEAD (`⊙` / `@`). */
  readonly headCommit: string;
  /** Alias of `commit` — merges share the commit node glyph. */
  readonly merge: string;
  /** Uncommitted working-tree marker (`○` / `o`). */
  readonly uncommitted: string;
  /** Stash side-leaf node in the gutter (`◇` / `s`) — spur off `stash^1`. */
  readonly stash: string;
  /** Checkout mark inside a branch chip (nf-fa-crosshairs / `+`). */
  readonly checkoutMark: string;
  /** Local↔remote sync mark inside a branch chip (nf-fa-exchange / `=`). */
  readonly syncMark: string;
  readonly vertical: string;
  readonly horizontal: string;
  readonly cornerDownRight: string; // ╮ — open lane to the right of node
  readonly cornerDownLeft: string; // ╭ — open lane to the left of node
  readonly cornerUpRight: string; // ╯ — close lane into node from the right
  readonly cornerUpLeft: string; // ╰ — close lane into node from the left
  readonly teeLeft: string;
  readonly teeRight: string;
  readonly teeDown: string;
  readonly teeUp: string;
  readonly cross: string;
};

const UNICODE: GraphGlyphSet = {
  commit: '●',
  headCommit: '⊙',
  merge: '●',
  uncommitted: '○',
  stash: '◇',
  // PUA Nerd Font icons — MesloLGM metrics stay 1-cell (misc unicode ⌖/⇄ bleed).
  checkoutMark: '', // nf-fa-crosshairs
  syncMark: '', // nf-fa-exchange
  vertical: '│',
  horizontal: '─',
  cornerDownRight: '╮',
  cornerDownLeft: '╭',
  cornerUpRight: '╯',
  cornerUpLeft: '╰',
  teeLeft: '┤',
  teeRight: '├',
  teeDown: '┬',
  teeUp: '┴',
  cross: '┼',
};

/** Classic `git log --graph` ASCII — single-column feel inside CELL_W pairs. */
const ASCII: GraphGlyphSet = {
  commit: '*',
  headCommit: '@',
  merge: '*',
  uncommitted: 'o',
  stash: 's',
  checkoutMark: '+',
  syncMark: '=',
  vertical: '|',
  horizontal: '-',
  cornerDownRight: '\\',
  cornerDownLeft: '/',
  cornerUpRight: '/',
  cornerUpLeft: '\\',
  teeLeft: '+',
  teeRight: '+',
  teeDown: '+',
  teeUp: '+',
  cross: '+',
};

/**
 * Active glyph set from `WS_STATUS_GLYPHS` (same flag as icons).
 *
 * Pass `ascii: true|false` to override the process env (tests).
 */
export function graphGlyphs(ascii: boolean = asciiGlyphs): GraphGlyphSet {
  return ascii ? ASCII : UNICODE;
}

File: crates/workspace-status-graph/src/glyphs.rs
//! Graph node, rail, and sync glyphs.
//!
//! Unicode matches `docs/git-graph-topology.md`.
//! ASCII matches `WS_STATUS_GLYPHS=ascii`. Topology is the same; only
//! the glyph map changes.

/// Columns occupied per logical lane (glyph + spacer / horizontal).
pub const CELL_W: usize = 2;

/// One paint mode for nodes, rails, and sync marks.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct GlyphSet {
    /// Regular commit node.
    pub commit: &'static str,
    /// HEAD commit node (including merge-at-HEAD).
    pub head_commit: &'static str,
    /// Uncommitted working-tree marker.
    pub uncommitted: &'static str,
    /// Stash side-leaf tip.
    pub stash: &'static str,
    /// Linked worktree marker. `ICON_LINKED_WORKTREE` (`` / `L`).
    pub worktree: &'static str,
    /// Ahead of upstream.
    pub ahead: &'static str,
    /// Behind upstream.
    pub behind: &'static str,
    /// Checkout mark on the named HEAD branch chip (`` / `+`).
    pub checkout_mark: &'static str,
    /// Synced local+remote mark inside a merged chip (`` / `=`).
    pub sync_mark: &'static str,
    /// Vertical through-rail.
    pub vertical: &'static str,
    /// Horizontal parent edge.
    pub horizontal: &'static str,
    /// Open lane to the right of the node (`╮` / `\`).
    pub corner_down_right: &'static str,
    /// Open lane to the left of the node (`╭` / `/`).
    pub corner_down_left: &'static str,
    /// Close lane into the node from the right (`╯` / `/`).
    pub corner_up_right: &'static str,
    /// Close lane into the node from the left (`╰` / `\`).
    pub corner_up_left: &'static str,
    /// Tee pointing left (`┤` / `+`).
    pub tee_left: &'static str,
    /// Tee pointing right (`├` / `+`).
    pub tee_right: &'static str,
    /// Tee pointing down (`┬` / `+`).
    pub tee_down: &'static str,
    /// Tee pointing up (`┴` / `+`).
    pub tee_up: &'static str,
    /// Four-way cross (`┼` / `+`).
    pub cross: &'static str,
}

/// Unicode glyphs used by the default widget.
pub const UNICODE: GlyphSet = GlyphSet {
    commit: "●",
    head_commit: "⊙",
    uncommitted: "○",
    stash: "◇",
    worktree: "",
    ahead: "↑",
    behind: "↓",
    checkout_mark: "",
    sync_mark: "",
    vertical: "│",
    horizontal: "─",
    corner_down_right: "╮",
    corner_down_left: "╭",
    corner_up_right: "╯",
    corner_up_left: "╰",
    tee_left: "┤",
    tee_right: "├",
    tee_down: "┬",
    tee_up: "┴",
    cross: "┼",
};

/// ASCII glyphs for terminals without the default font.
pub const ASCII: GlyphSet = GlyphSet {
    commit: "*",
    head_commit: "@",
    uncommitted: "o",
    stash: "s",
    worktree: "L",
    ahead: "^",
    behind: "v",
    checkout_mark: "+",
    sync_mark: "=",
    vertical: "|",
    horizontal: "-",
    corner_down_right: "\\",
    corner_down_left: "/",
    corner_up_right: "/",
    corner_up_left: "\\",
    tee_left: "+",
    tee_right: "+",
    tee_down: "+",
    tee_up: "+",
    cross: "+",
};

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn worktree_glyph_is_linked_not_emoji() {
        assert_eq!(UNICODE.worktree, "");
        assert_eq!(ASCII.worktree, "L");
        assert_ne!(UNICODE.worktree, "🔗");
        assert_ne!(ASCII.worktree, "wt");
    }
}
